x-pack/plugins/security/server/session_management/session.ts

```diff
@@ -0,0 +1,405 @@
+    this.idleTimeout = this.options.config.session.idleTimeout;
+    this.lifespan = this.options.config.session.lifespan;
+
+    // The timeout after which we update index is two times longer than configured idle timeout
+    // since index updates are costly and we want to minimize them.
+    this.idleIndexUpdateTimeout = this.options.config.session.idleTimeout
+      ? this.options.config.session.idleTimeout.asMilliseconds() * 2
+      : null;
+  }
+
+  /**
+   * Extracts session value for the specified request. Under the hood it can clear session if it is
+   * invalid or created by the legacy versions of Kibana.
+   * @param request Request instance to get session value for.
+   */
+  async get(request: KibanaRequest) {
+    const sessionCookieValue = await this.options.sessionCookie.get(request);
+    if (!sessionCookieValue) {
+      return null;
+    }
+
+    const sessionLogger = this.getLoggerForSID(sessionCookieValue.sid);
+    const now = Date.now();
+    if (
+      (sessionCookieValue.idleTimeoutExpiration &&
+        sessionCookieValue.idleTimeoutExpiration < now) ||
+      (sessionCookieValue.lifespanExpiration && sessionCookieValue.lifespanExpiration < now)
+    ) {
+      sessionLogger.debug('Session has expired and will be invalidated.');
+      await this.clear(request);
+      return null;
+    }
+
+    const sessionIndexValue = await this.options.sessionIndex.get(sessionCookieValue.sid);
+    if (!sessionIndexValue) {
+      sessionLogger.debug(
+        'Session value is not available in the index, session cookie will be invalidated.'
+      );
+      await this.options.sessionCookie.clear(request);
+      return null;
+    }
+
+    let decryptedContent: SessionValueContentToEncrypt;
+    try {
+      decryptedContent = JSON.parse(
+        (await this.crypto.decrypt(sessionIndexValue.content, sessionCookieValue.aad)) as string
+      );
+    } catch (err) {
+      sessionLogger.warn(
+        `Unable to decrypt session content, session will be invalidated: ${err.message}`
+      );
+      await this.clear(request);
+      return null;
+    }
+
+    return {
+      ...Session.sessionIndexValueToSessionValue(sessionIndexValue, decryptedContent),
+      // Unlike session index, session cookie contains the most up to date idle timeout expiration.
+      idleTimeoutExpiration: sessionCookieValue.idleTimeoutExpiration,
+    };
+  }
+
+  /**
+   * Creates new session document in the session index encrypting sensitive state.
+   * @param request Request instance to create session value for.
+   * @param sessionValue Session value parameters.
+   */
+  async create(
+    request: KibanaRequest,
+    sessionValue: Readonly<
+      Omit<SessionValue, 'sid' | 'idleTimeoutExpiration' | 'lifespanExpiration' | 'metadata'>
+    >
+  ) {
+    const [sid, aad] = await Promise.all([
+      this.randomBytes(32).then((sidBuffer) => sidBuffer.toString('base64')),
+      this.randomBytes(32).then((aadBuffer) => aadBuffer.toString('base64')),
```

Comment:
Nit: we should probably define constants like `SID_BYTE_LENGTH` and `AAD_BYTE_LENGTH` at the top of the file and use them here so there's no ambiguity in our approach. Should also probably include a comment that indicates we have determined a 256-bit SID/AAD to be safe to use.

Reply:
Yep, good point!